allennlp/modules/conditional_random_field_lannoy.py

```diff
@@ -0,0 +1,486 @@
+    """
+
+    if to_tag == "START" or from_tag == "END":
+        # Cannot transition into START or from END
+        return False
+
+    if constraint_type == "BIOUL":
+        if from_tag == "START":
+            return to_tag in ("O", "B", "U")
+        if to_tag == "END":
+            return from_tag in ("O", "L", "U")
+        return any(
+            [
+                # O can transition to O, B-* or U-*
+                # L-x can transition to O, B-*, or U-*
+                # U-x can transition to O, B-*, or U-*
+                from_tag in ("O", "L", "U") and to_tag in ("O", "B", "U"),
+                # B-x can only transition to I-x or L-x
+                # I-x can only transition to I-x or L-x
+                from_tag in ("B", "I") and to_tag in ("I", "L") and from_entity == to_entity,
+            ]
+        )
+    elif constraint_type == "BIO":
+        if from_tag == "START":
+            return to_tag in ("O", "B")
+        if to_tag == "END":
+            return from_tag in ("O", "B", "I")
+        return any(
+            [
+                # Can always transition to O or B-x
+                to_tag in ("O", "B"),
+                # Can only transition to I-x from B-x or I-x
+                to_tag == "I" and from_tag in ("B", "I") and from_entity == to_entity,
+            ]
+        )
+    elif constraint_type == "IOB1":
+        if from_tag == "START":
+            return to_tag in ("O", "I")
+        if to_tag == "END":
+            return from_tag in ("O", "B", "I")
+        return any(
+            [
+                # Can always transition to O or I-x
+                to_tag in ("O", "I"),
+                # Can only transition to B-x from B-x or I-x, where
+                # x is the same tag.
+                to_tag == "B" and from_tag in ("B", "I") and from_entity == to_entity,
+            ]
+        )
+    elif constraint_type == "BMES":
+        if from_tag == "START":
+            return to_tag in ("B", "S")
+        if to_tag == "END":
+            return from_tag in ("E", "S")
+        return any(
+            [
+                # Can only transition to B or S from E or S.
+                to_tag in ("B", "S") and from_tag in ("E", "S"),
+                # Can only transition to M-x from B-x, where
+                # x is the same tag.
+                to_tag == "M" and from_tag in ("B", "M") and from_entity == to_entity,
+                # Can only transition to E-x from B-x or M-x, where
+                # x is the same tag.
+                to_tag == "E" and from_tag in ("B", "M") and from_entity == to_entity,
+            ]
+        )
+    else:
+        raise ConfigurationError(f"Unknown constraint type: {constraint_type}")
+
+
+class ConditionalRandomFieldLannoy(torch.nn.Module):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
+
```

Comment:
Also would be good to have a note about the weight strategy here.

Reply:
Done!